ferretdb/ferretdb.go

```diff
@@ -214,21 +223,23 @@ func (f *FerretDB) MongoDBURI() string {
 	return u.String()
 }
 
-// logger is a global logger used by FerretDB.
-//
-// TODO https://github.com/FerretDB/FerretDB/issues/4014
-var logger *zap.Logger
+var (
+	loggerOnce sync.Once
+	logger     *zap.Logger
+)
 
-// Initialize the global logger there to avoid creating too many issues for zap users that initialize it in their
-// `main()` functions. It is still not a full solution; eventually, we should remove the usage of the global logger.
-//
-// TODO https://github.com/FerretDB/FerretDB/issues/4014
-func init() {
-	l := zap.ErrorLevel
-	if version.Get().DebugBuild {
-		l = zap.DebugLevel
-	}
+// getGlobalLogger retrieves or creates a global logger using
+// a loggerOnce to ensure it is created only once.
+func getGlobalLogger() *zap.Logger {
+	loggerOnce.Do(func() {
+		level := zap.ErrorLevel
+		if version.Get().DebugBuild {
+			level = zap.DebugLevel
+		}
+
+		logging.Setup(level, "console", "")
```

Comment:
`logging.Setup` also adds a zap hook for `RecentEntries`. If we don't call this method, then `getLog` command handler will not work correctly